tools/rpkg/src/duckdbr.cpp

```diff
@@ -1169,6 +1170,113 @@ SEXP duckdb_unregister_R(SEXP connsexp, SEXP namesexp) {
 	return R_NilValue;
 }
 
+class RArrowTabularStreamFactory {
+public:
+	RArrowTabularStreamFactory(SEXP export_fun_p, SEXP arrow_tabular_p)
+	    : arrow_tabular(arrow_tabular_p), export_fun(export_fun_p) {};
+	static unique_ptr<ArrowArrayStreamWrapper> Produce(uintptr_t factory_p) {
+		RProtector r;
+		int err;
+		auto factory = (RArrowTabularStreamFactory *)factory_p;
+
+		// export the arrow tabular to a stream pointer
+		auto export_call = r.Protect(Rf_lang2(factory->export_fun, factory->arrow_tabular));
+		SEXP stream_ptr_sexp = r.Protect(R_tryEval(export_call, R_GlobalEnv, &err));
+		if (err) {
+			Rf_error("Failed to produce arrow stream");
+		}
+		// TODO this stream pointer leaks?
+		// TODO this may fail on 32 bit
+		auto stream_ptr = (void *)(uintptr_t)REAL(stream_ptr_sexp)[0];
+
+		auto res = make_unique<ArrowArrayStreamWrapper>();
+		res->arrow_array_stream = *(ArrowArrayStream *)stream_ptr;
+		return res;
+	}
+
+	SEXP arrow_tabular;
+	SEXP export_fun;
+};
+
+static SEXP duckdb_finalize_arrow_factory_R(SEXP factorysexp) {
+	if (TYPEOF(factorysexp) != EXTPTRSXP) {
+		Rf_error("duckdb_finalize_arrow_factory_R: Need external pointer parameter");
+	}
+	auto *factoryaddr = (RArrowTabularStreamFactory *)R_ExternalPtrAddr(factorysexp);
+	if (factoryaddr) {
+		R_ClearExternalPtr(factorysexp);
+		delete factoryaddr;
+	}
+	return R_NilValue;
+}
+
+SEXP duckdb_register_arrow_R(SEXP connsexp, SEXP namesexp, SEXP funsexp, SEXP valuesexp) {
+	if (TYPEOF(connsexp) != EXTPTRSXP) {
+		Rf_error("duckdb_register_R: Need external pointer parameter for connection");
+	}
+	Connection *conn = (Connection *)R_ExternalPtrAddr(connsexp);
+	if (!conn) {
+		Rf_error("duckdb_register_R: Invalid connection");
+	}
+	if (TYPEOF(namesexp) != STRSXP || Rf_length(namesexp) != 1) {
+		Rf_error("duckdb_register_R: Need single string parameter for name");
+	}
+	auto name = string(CHAR(STRING_ELT(namesexp, 0)));
+	if (name.empty()) {
+		Rf_error("duckdb_register_R: name parameter cannot be empty");
+	}
+
+	if (!Rf_isFunction(funsexp)) {
+		Rf_error("duckdb_register_R: Need single function parameter");
+	}
+
+	if (!Rf_inherits(valuesexp, "ArrowTabular")) {
+		Rf_error("duckdb_register_R: Need ArrowTabular object as value parameter");
+	}
```

Comment:
I don't think you want this. This would exclude the Arrow classes `Dataset`, `arrow_dplyr_query`, etc. I think it would be better to let `Scanner$create()` (called in your `export_fun()`) [handle its validation](https://github.com/apache/arrow/blob/master/r/R/dataset-scan.R#L74-L92) and error if it gets something it doesn't know how to handle.

Reply:
Yes, great, removed it there.